File: bro-agent-api/src/main/java/com/ericsson/adp/mgmt/bro/api/agent/WaitingForActionState.java
/*------------------------------------------------------------------------------
 *******************************************************************************
 * COPYRIGHT Ericsson 2019
 *
 * The copyright to the computer program(s) herein is the property of
 * Ericsson Inc. The programs may be used and/or copied only with written
 * permission from Ericsson Inc. or in accordance with the terms and
 * conditions stipulated in the agreement/contract under which the
 * program(s) have been supplied.
 *******************************************************************************
 *----------------------------------------------------------------------------*/
package com.ericsson.adp.mgmt.bro.api.agent;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import com.ericsson.adp.mgmt.control.OrchestratorControl;
import com.ericsson.adp.mgmt.control.OrchestratorMessageType;

/**
 * Waits for Orchestrator control message.
 */
public class WaitingForActionState extends AgentState {

    private static final Logger log = LogManager.getLogger(WaitingForActionState.class);

    /**
     * Creates state.
     * @param agent The Agent waiting for action.
     */
    public WaitingForActionState(final Agent agent) {
        super(agent);
    }

    @Override
    public AgentState processMessage(final OrchestratorControl message) {
        if (message.hasRegisterAcknowledge()) {
            log.info(message);
            return this;
        } else if (isPreparationMessage(message)) {
            switch (message.getAction()) {
                case BACKUP:
                    log.info("The Orchestrator has requested that the Agent performs a backup: {}", message.getPreparation().getBackupName());
                    return new BackupPreparationState(this.agent, new ActionInformation(message.getPreparation()));
                case RESTORE:
                    log.info("The Orchestrator has requested that the Agent restores backup: {}", message.getPreparation().getBackupName());
                    return new RestorePreparationState(this.agent, message.getPreparation());
            }
        }
        log.info("Unexpected message received <{}>", message);
        return this;

    }

    private boolean isPreparationMessage(final OrchestratorControl message) {
        return message.getOrchestratorMessageType().equals(OrchestratorMessageType.PREPARATION) && message.hasPreparation();
    }

    @Override
    public void trigger() {
        // Do nothing
    }

    @Override
    public AgentState finishAction() {
        return this;
    }

    @Override
    protected AgentState cancelAction() {
        return this;
    }

}

File: bro-agent-api/src/main/python/bro_agent/agent/WaitingForActionState.py
import logging
import bro_agent.agent.states

from bro_agent.generated.OrchestratorControl_pb2 import OrchestratorMessageType
from bro_agent.generated.Action_pb2 import Action


class WaitingForActionState(bro_agent.agent.AgentState.AgentState):
    """ Waits for Orchestrator control message """

    def __init__(self, agent):
        # @param agent The Agent waiting for action
        bro_agent.agent.AgentState.AgentState.__init__(self, agent)

    @staticmethod
    def is_preparation_message(message):
        """ Is the message a Preparation message? """
        return (message.orchestratorMessageType == OrchestratorMessageType.PREPARATION) and (message.preparation is not None)

    def process_message(self, message):
        if self.should_start_backup(message):
            logging.info("The Orchestrator has requested that the Agent creates backup: %s", message.preparation.backupName)
            return bro_agent.agent.BackupExecutionState.BackupExecutionState(self.agent, message.preparation)
        if self.should_start_restore(message):
            logging.info("The Orchestrator has requested that the Agent restores backup: %s", message.preparation.backupName)
            return bro_agent.agent.RestorePreparationState.RestorePreparationState(self.agent, message.preparation)
        logging.warning("The Orchestrator has requested an unknown operation: %s", message.action)
        return self

    def should_start_backup(self, message):
        return Action.BACKUP == message.action and self.is_preparation_message(message)

    def should_start_restore(self, message):
        return Action.RESTORE == message.action and self.is_preparation_message(message)

    def trigger(self):
        pass

    def finish_action(self):
        return self

    def cancel_action(self):
        return self
